### ISSUE TITLE
Operator fails on Juju 2.7.0 and below

### ISSUE THREAD

**Issue by chris-sanders** [canonical/operator#223]
I've deployed my first operator charm and found that on my 2.7.0 controller the framework throws an exception because it tries to use "relation-get" with the "--app" flag which is not supported on 2.7.0. 

Operator should really put a try statement around the relation subprocesses and simply ignore application level data on versions of Juju that haven't implemented it.

**Comment by niemeyer** [comment]
Indeed, and thanks for the report. We already have logic to test for the juju version, so it should be easy to just skip it or return values as unset in those cases.

**Comment by jameinel** [comment]
Are you sure this was with 2.7.0 ? Because when I test it:
```
$ juju status -m lxd-2.7.0:default
Model    Controller  Cloud/Region         Version  SLA          Timestamp
default  lxd-2.7.0   localhost/localhost  2.7.0    unsupported  14:36:25+04:00

App              Version  Status   Scale  Charm            Store  Rev  OS      Notes
dummy-sink                waiting      1  dummy-sink       local    0  ubuntu
dummy-source              blocked      1  dummy-source     local    0  ubuntu

$ juju run -m lxd-2.7.0:default --unit dummy-sink/0 'relation-get -r 0 --app - dummy-source/0'
{}
$ juju run -m lxd-2.7.0:default --unit dummy-source/0 'relation-set -r 0 --app foo=bar'
$ juju run -m lxd-2.7.0:default --unit dummy-sink/0 'relation-get -r 0 --app - dummy-source/0'
foo: bar
```

It seems to be supported just fine in 2.7.0 (there may be bugs, which necessitated 2.7.1+ to be released, but relation-get --app was there.)
Do you have more details on what actually failed?

**Comment by jameinel** [comment]
Sorry, hit the wrong button.

I do remember this bug:
https://bugs.launchpad.net/juju/+bug/1854348

Which was flagged as fixed in 2.7.3 not 2.7.4.  And I do get this failure on 2.7.0:
```
$ juju run -m lxd-2.7.0:default --unit dummy-sink/0 'echo `is-leader`; relation-get -r 0 --app - dummy-sink/0'
True
ERROR permission denied
$ juju run -m lxd-2.7.0:default --unit dummy-sink/0 'echo `is-leader`; relation-get -r 0 --app - dummy-source/0'
True
foo: bar
```

Which says that you aren't able to read your own data bag, but you are able to read the remote data bag.
My concern with "we're trying to deploy this charm but it is failing", if it is expecting data to be somewhere, but that data isn't available, even if we paper over the "permission denied", that still doesn't put the data they need to run into the place they are looking for it.

So it depends if it is a relation-get call that the charm is depending upon, or whether it is a 'casual' use-it-if-it-is-available.

**Comment by chris-sanders** [comment]
This bug is just about Operator trying to use --app on versions that don't support it.

The following throws an error if juju doesn't support --app and means that a cloud running a controller before the 2.7 series can not have an operator charm installed on it.
`
def on_my_relation_changed(self, event):
    """Handle relation change."""
    self.state.data = event.relation.data[event.unit]
`
There might be more to consider with other use cases, but the use case I'm describing is simply an operator charm, not using app data at all, just legacy relations can't be installed on stable 2.5 and 2.6 controllers.
